File: mqtt-client-Android/app/src/main/java/io/emqx/mqtt/MessageRecyclerViewAdapter.java
package io.emqx.mqtt;

import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;


public class MessageRecyclerViewAdapter extends RecyclerView.Adapter<MessageRecyclerViewAdapter.ViewHolder> {

    private final List<Message> mValues;

    public MessageRecyclerViewAdapter(List<Message> items) {
        mValues = items;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_message, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.mItem = mValues.get(position);
        holder.mTopicView.setText("Topic: " + mValues.get(position).getTopic());
        holder.mContentView.setText("Payload: " + mValues.get(position).getMessage().toString());
    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final View mView;
        public final TextView mTopicView;
        public final TextView mContentView;
        public Message mItem;

        public ViewHolder(View view) {
            super(view);
            mView = view;
            mTopicView = view.findViewById(R.id.tv_topic);
            mContentView = view.findViewById(R.id.tv_payload);
        }

        @Override
        public String toString() {
            return super.toString() + " '" + mContentView.getText() + "'";
        }
    }
}

File: mqtt-client-Android-Kotlin/app/src/main/java/io/emqx/mqtt/MessageRecyclerViewAdapter.kt
package io.emqx.mqtt

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class MessageRecyclerViewAdapter(private val mValues: List<Message>) :
    RecyclerView.Adapter<MessageRecyclerViewAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_message, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.mItem = mValues[position]
        holder.mTopicView.text = "Topic: " + mValues[position].topic
        holder.mContentView.text = "Payload: " + mValues[position].message.toString()
    }

    override fun getItemCount(): Int {
        return mValues.size
    }

    inner class ViewHolder(mView: View) : RecyclerView.ViewHolder(
        mView
    ) {
        val mTopicView: TextView = mView.findViewById(R.id.tv_topic)
        val mContentView: TextView = mView.findViewById(R.id.tv_payload)
        var mItem: Message? = null
        override fun toString(): String {
            return super.toString() + " '" + mContentView.text + "'"
        }

    }
}